src/transformers.jl

```diff
@@ -0,0 +1,66 @@
+
+"""
+    Transformation
+
+Abstract supertype for all transformations.
+"""
+abstract type Transformation end
+
+# Make Transforms callable types
+(t::Transformation)(x; kwargs...) = transform(x, t; kwargs...)
+
+"""
+    transform!(data::T, transformation::Transformation; kwargs...) -> T
+
+Apply `transformation` to the `data` in-place.
+Where possible, this should be extended for new data types `T`.
+"""
+function transform! end
+
+"""
+    transform(data::T, transformation::Transformation; kwargs...) -> T
+
+Non-in-place version of [`transform!`](@ref), which it delegates to by default.
+Does not need to be extended unless in-place transformation is not possible.
+"""
+function transform end
```

Comment:
Yeah I'm not sure...

On the one had we're kinda stuck with `transform` for the sake of consistency in our downstream dependencies (Forecasters).

But on the other hand, `transform` doesn't make sense in the context of `Imputor`s and `apply` would be better there. 

I wonder is there a way to distinguish how these methods/terms apply in certain contexts that makes intuitive sense?
e.g. you `apply` a Transform/Imputor to a certain data type, but you should overload `transform` when combining these operations in a pipeline?

So that `apply(data, ::Union{Transform, Imputor})` make sense but `apply(::MyModel, data)` doesn't. Conversely, `transform(data, ::Union{Transform, Imputor})` doesn't make (total) sense but `transform(::MyModel, data)` does?

Reply:
That distinction of `transform`/`apply` for model/data sounds reasonable to me Glenn.

Is it the case that we've made Transforms callable types here, but this is not so for Imputors? I'm wondering if the `transform`/`apply` distinction matters much externally if `transform` is almost always used via the callable type.

Reply:
Yeah I think it would, because you'd want Impute to extend the Transforms API, which is only possible if they use the same function underneath.

Reply:
I'm confused, so what would Impute.jl be extending `transform` or `apply`? NOTE: If the API was changed to `transform(MyType, data)` then I'd need to update and deprecate almost every method in Impute.jl

Reply:
Sorry, I'm also working this out as I type it.

>Impute.jl be extending transform or apply?

Impute would extend `apply`.

> If the API was changed to `transform(MyType, data)` then I'd need to update and deprecate almost every method in Impute.jl

Sorry, I could have been clearer. I'm not suggesting we change the API to this. 
I was just using that example (with `MyModel`) to showcase how we use currently use (and overload) `transform`, which is owned by the Forecasters API.

So if we want `Transform` to also own the `transform` method as well (which I'm guessing it should), then I'm trying to use our current application of it as an example of how we might want to distinguish it from `apply`.

e.g.
`Transform.apply` is a (potentially private?) method that is overloaded for new `Transform`/`Imputor`s.
`Transform.transform` is just a public interface method that is overloaded on types that want to perform feature engineering using `Transform`s/`Imputor`s.

Note: I'm possibly trying too hard to shoe-horn in both `apply` and `transform` into this package so let me know if this doesn't make much sense. Especially when trying to justify `transform` my basis is that:
1. We are using a method with that name already and so it should belong somewhere.  
b) It may as well be owned by `Transforms` because we can then define/enforce its contract at this higher level rather than in a dependent package like Forecaster (which isn't the only potential use-case in the long-term).

Does this make sense?

Reply:
I've renamed `transform`->`apply` and added _new_ methods for `transform` with docstrings that I hope explain what I'm getting at.

Reply:
> Note: I'm possibly trying too hard to shoe-horn in both apply and transform into this package so let me know if this doesn't make much sense. Especially when trying to justify transform my basis is that:
>
>    We are using a method with that name already and so it should belong somewhere.
>    b) It may as well be owned by Transforms because we can then define/enforce its contract at this higher level rather than in a dependent package like Forecaster (which isn't the only potential use-case in the long-term).

It does feel somewhat awkward that the package is handling both individual data transformations and feature engineering pipelines (though the latter only in abstract). It seems like the new role of `transform` is just a contract and won't get more complicated within this package, right? If that's the case I think the awkwardness is outweighed by the reasons you give.

Reply:
Yeah, we may find that we can generalize things such that we don't need both in the future,  but that distinction makes sense to me. Anyways, thanks for clarify, this should be pretty easy to extend in Impute.jl.

Reply:
> It seems like the new role of transform is just a contract and won't get more complicated within this package, right?

yes this is correct - there might be certain interface tests we'll eventually write but for now it just says "Transforms owns the transform method".